modules/table/src/main/java/org/apache/ignite/internal/table/distributed/PartitionReplicatorNodeRecovery.java

```diff
@@ -0,0 +1,195 @@
+            IntFunction<TableImpl> tableSupplier) {
+        this.metaStorageManager = metaStorageManager;
+        this.messagingService = messagingService;
+        this.clusterNodeResolver = clusterNodeResolver;
+        this.tableSupplier = tableSupplier;
+    }
+
+    /**
+     * Starts the component.
+     */
+    void start() {
+        addMessageHandler();
+    }
+
+    private void addMessageHandler() {
+        messagingService.addMessageHandler(TableMessageGroup.class, (message, sender, correlationId) -> {
+            if (message instanceof HasDataRequest) {
+                // This message queries if a node has any data for a specific partition of a table
+                assert correlationId != null;
+
+                HasDataRequest msg = (HasDataRequest) message;
+
+                int tableId = msg.tableId();
+                int partitionId = msg.partitionId();
+
+                boolean contains = false;
+
+                TableImpl table = tableSupplier.apply(tableId);
+
+                if (table != null) {
+                    MvTableStorage storage = table.internalTable().storage();
+
+                    MvPartitionStorage mvPartition = storage.getMvPartition(partitionId);
+
+                    // If node's recovery process is incomplete (no partition storage), then we consider this node's
+                    // partition storage empty.
+                    if (mvPartition != null) {
+                        contains = mvPartition.closestRowId(RowId.lowestRowId(partitionId)) != null;
+                    }
+                }
+
+                messagingService.respond(sender, TABLE_MESSAGES_FACTORY.hasDataResponse().result(contains).build(), correlationId);
+            }
+        });
+    }
+
+    /**
+     * Returns a future that completes with a decision: should we start the corresponding group locally or not.
+     *
+     * @param tablePartitionId ID of the table partition.
+     * @param internalTable Table we are working with.
+     * @param newConfiguration New configuration that is going to be applied if we'll start the group.
+     * @param localMemberAssignment Assignment of this node in this group.
+     */
+    CompletableFuture<Boolean> shouldStartGroup(
+            TablePartitionId tablePartitionId,
+            InternalTable internalTable,
+            PeersAndLearners newConfiguration,
+            Assignment localMemberAssignment
+    ) {
+        int tableId = tablePartitionId.tableId();
+        int partId = tablePartitionId.partitionId();
+
+        // If Raft is running in in-memory mode or the PDS has been cleared, we need to remove the current node
+        // from the Raft group in order to avoid the double vote problem.
+        // <MUTED> See https://issues.apache.org/jira/browse/IGNITE-16668 for details.
+        // TODO: https://issues.apache.org/jira/browse/IGNITE-19046 Restore "|| !hasData"
+        if (internalTable.storage().isVolatile()) {
```

Comment:
I would suggest to invert the `if` condition, so that we will have less indentation

Reply:
Inverting an IF would require to change the comment, I preferred to extract a method, the code looks simpler and is less indented